File: app/src/test/java/com/aus/ausgegeben/data/FirestorePathsTest.kt
package com.aus.ausgegeben.data

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * DEL-1 regression: a new account-owned Firestore path must be impossible to add
 * without an explicit decision about whether account deletion covers it. Mirrors
 * web/src/repositories/firestorePaths.test.ts — keep both in sync.
 */
class FirestorePathsTest {

    @Test
    fun `classifies every whole-collection constant as deletable`() {
        assertTrue(FirestorePaths.CATEGORIES_COLLECTION in FirestorePaths.DELETABLE_USER_COLLECTIONS)
        assertTrue(FirestorePaths.EXPENSES_COLLECTION in FirestorePaths.DELETABLE_USER_COLLECTIONS)
    }

    @Test
    fun `classifies every named single-document constant as exactly one of deletable or intentionally retained`() {
        val namedDocs = listOf(
            FirestorePaths.UserDocPath(FirestorePaths.SETTINGS_COLLECTION, FirestorePaths.PREFERENCES_DOC),
            FirestorePaths.UserDocPath(FirestorePaths.META_COLLECTION, FirestorePaths.DEDUPE_DOC),
            FirestorePaths.UserDocPath(FirestorePaths.META_COLLECTION, FirestorePaths.ACCOUNT_DELETION_DOC),
        )
        for (doc in namedDocs) {
            val deletable = doc in FirestorePaths.DELETABLE_USER_DOCS
            val retained = doc in FirestorePaths.INTENTIONALLY_RETAINED_USER_DOCS
            assertTrue(
                "$doc must be classified as exactly one of deletable/retained (deletable=$deletable, retained=$retained)",
                deletable != retained,
            )
        }
    }

    /**
     * The two tests above only check a fixed, hand-written set of names they already
     * expect to see — they would not fail if a *new* `const val` were added to
     * [FirestorePaths] without ever being added to a classification list (independently
     * proven during review, on the mirrored web module: adding an unclassified
     * `RECEIPTS_COLLECTION` constant left both prior-style tests green). This test
     * instead reflects over the object's own declared String constants, so it fails on
     * any future constant no matter what it's called.
     */
    @Test
    fun `every declared String path constant is reachable through a classification list`() {
        val classifiedNames = buildSet {
            addAll(FirestorePaths.DELETABLE_USER_COLLECTIONS)
            FirestorePaths.DELETABLE_USER_DOCS.forEach { add(it.collection); add(it.id) }
            FirestorePaths.INTENTIONALLY_RETAINED_USER_DOCS.forEach { add(it.collection); add(it.id) }
        }

        val stringConstants = FirestorePaths::class.java.declaredFields
            .filter { java.lang.reflect.Modifier.isStatic(it.modifiers) && it.type == String::class.java }
            .associate { field ->
                field.isAccessible = true
                field.name to (field.get(null) as String)
            }
        assertTrue("expected at least one String constant on FirestorePaths", stringConstants.isNotEmpty())

        for ((name, value) in stringConstants) {
            assertTrue(
                "$name = \"$value\" is declared but does not appear in DELETABLE_USER_COLLECTIONS, " +
                    "DELETABLE_USER_DOCS, or INTENTIONALLY_RETAINED_USER_DOCS — classify it.",
                value in classifiedNames,
            )
        }
    }

    /**
     * The registry can be internally consistent yet still miss a path that
     * firestore.rules actually permits — this ties it back to the authoritative schema
     * definition instead of only checking the registry against itself.
     */
    @Test
    fun `covers every per-user collection and named document that firestore rules actually permits`() {
        val rulesFile = findFirestoreRules()
        val rules = rulesFile.readText()

        val usersBlockStart = rules.indexOf("match /users/{userId}")
        assertTrue("firestore.rules must declare match /users/{userId}", usersBlockStart >= 0)
        val body = rules.substring(usersBlockStart)

        val collectionRegex = Regex("""match /(\w+)/\{(\w+)\}""")
        val subcollections = collectionRegex.findAll(body)
            .map { it.groupValues[1] }
            .filter { it != "users" }
            .toSet()

        assertEquals(
            setOf(
                FirestorePaths.CATEGORIES_COLLECTION,
                FirestorePaths.EXPENSES_COLLECTION,
                FirestorePaths.SETTINGS_COLLECTION,
                FirestorePaths.META_COLLECTION,
            ),
            subcollections,
        )

        val docIdRegex = Regex("""docId == '([\w-]+)'""")
        val docIdLiterals = docIdRegex.findAll(rules).map { it.groupValues[1] }.toSet()

        assertEquals(
            setOf(
                FirestorePaths.PREFERENCES_DOC,
                FirestorePaths.DEDUPE_DOC,
                FirestorePaths.ACCOUNT_DELETION_DOC,
            ),
            docIdLiterals,
        )
    }

    /** Gradle's test working directory is the module dir (app/); walk up to find the repo root. */
    private fun findFirestoreRules(): File {
        var dir = File(".").absoluteFile
        repeat(6) {
            val candidate = File(dir, "firestore.rules")
            if (candidate.exists()) return candidate
            dir = dir.parentFile ?: return@repeat
        }
        throw IllegalStateException("Could not locate firestore.rules from ${File(".").absolutePath}")
    }
}

File: web/src/repositories/firestorePaths.test.ts
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';
import * as firestorePaths from './firestorePaths';
import {
  ACCOUNT_DELETION_DOC,
  CATEGORIES_COLLECTION,
  DEDUPE_DOC,
  DELETABLE_USER_COLLECTIONS,
  DELETABLE_USER_DOCS,
  EXPENSES_COLLECTION,
  INTENTIONALLY_RETAINED_USER_DOCS,
  META_COLLECTION,
  PREFERENCES_DOC,
  SETTINGS_COLLECTION,
} from './firestorePaths';

/**
 * DEL-1 regression: a new account-owned Firestore path must be impossible to add
 * without an explicit decision about whether account deletion covers it.
 */
describe('firestorePaths registry (DEL-1)', () => {
  it('classifies every whole-collection constant as deletable', () => {
    // categories/ and expenses/ are deleted document-by-document (not as single
    // named docs), so they are declared directly rather than as {collection, id} pairs.
    expect(DELETABLE_USER_COLLECTIONS).toContain(CATEGORIES_COLLECTION);
    expect(DELETABLE_USER_COLLECTIONS).toContain(EXPENSES_COLLECTION);
  });

  it('classifies every named single-document constant as exactly one of deletable or intentionally retained', () => {
    const namedDocs: Array<{ collection: string; id: string; label: string }> = [
      { collection: SETTINGS_COLLECTION, id: PREFERENCES_DOC, label: 'settings/preferences' },
      { collection: META_COLLECTION, id: DEDUPE_DOC, label: 'meta/dedupe' },
      { collection: META_COLLECTION, id: ACCOUNT_DELETION_DOC, label: 'meta/accountDeletion' },
    ];

    for (const doc of namedDocs) {
      const deletable = DELETABLE_USER_DOCS.some((d) => d.collection === doc.collection && d.id === doc.id);
      const retained = INTENTIONALLY_RETAINED_USER_DOCS.some(
        (d) => d.collection === doc.collection && d.id === doc.id,
      );
      expect(
        deletable !== retained, // exactly one, never both, never neither
        `${doc.label} must be classified as exactly one of deletable/retained (deletable=${deletable}, retained=${retained})`,
      ).toBe(true);
    }
  });

  // The two tests above only check a fixed, hand-written set of names they already
  // expect to see — they would not fail if a *new* constant were added to the module
  // without ever being added to a classification list (independently proven during
  // review: adding an unclassified `export const RECEIPTS_COLLECTION = 'receipts'`
  // above left both prior tests green). This test instead enumerates the module's
  // actual exports, so it fails on any future constant no matter what it's called.
  it('every exported string constant is reachable through a classification list', () => {
    const classifiedNames = new Set<string>([
      ...DELETABLE_USER_COLLECTIONS,
      ...DELETABLE_USER_DOCS.flatMap((d) => [d.collection, d.id]),
      ...INTENTIONALLY_RETAINED_USER_DOCS.flatMap((d) => [d.collection, d.id]),
    ]);

    const stringExports = Object.entries(firestorePaths as Record<string, unknown>).filter(
      (entry): entry is [string, string] => typeof entry[1] === 'string',
    );
    // Sanity check on the test itself: if this module ever stopped exporting any bare
    // string constant at all, the loop below would vacuously pass without checking
    // anything.
    expect(stringExports.length).toBeGreaterThan(0);

    for (const [name, value] of stringExports) {
      expect(
        classifiedNames.has(value),
        `${name} = '${value}' is exported but does not appear in DELETABLE_USER_COLLECTIONS, ` +
          `DELETABLE_USER_DOCS, or INTENTIONALLY_RETAINED_USER_DOCS — classify it.`,
      ).toBe(true);
    }
  });

  // The registry can be internally consistent yet still miss a path that firestore.rules
  // actually permits — this ties it back to the authoritative schema definition instead
  // of only checking the registry against itself.
  it('covers every per-user collection and named document that firestore.rules actually permits', () => {
    const rulesPath = path.resolve(__dirname, '../../../firestore.rules');
    const rules = readFileSync(rulesPath, 'utf8');

    // Every `match /<segment>/{...}` nested directly under `match /users/{userId} { ... }`.
    const usersBlockStart = rules.indexOf('match /users/{userId}');
    expect(usersBlockStart).toBeGreaterThan(-1);
    const collectionMatches = [
      ...rules.slice(usersBlockStart).matchAll(/match \/(\w+)\/\{(\w+)\}/g),
    ].map((m) => m[1]);
    // The outer match itself isn't a subcollection segment.
    const subcollections = collectionMatches.filter((name) => name !== 'users');

    expect(new Set(subcollections)).toEqual(
      new Set([CATEGORIES_COLLECTION, EXPENSES_COLLECTION, SETTINGS_COLLECTION, META_COLLECTION]),
    );

    // Every `docId == '...'` literal anywhere in the file — these are the only doc ids
    // the rules ever grant any permission for under settings/ or meta/.
    const docIdLiterals = new Set(
      [...rules.matchAll(/docId == '([\w-]+)'/g)].map((m) => m[1]),
    );
    expect(docIdLiterals).toEqual(new Set([PREFERENCES_DOC, DEDUPE_DOC, ACCOUNT_DELETION_DOC]));
  });
});
